packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java

```diff
@@ -0,0 +1,147 @@
+    Assert.assertFalse(updatesToDelete.contains(updateRollout2));
+  }
+
+  @Test
+  public void testSelectUpdatesToDelete_NoneOlderMatching() {
+    // if there is no older update that matches the manifest filters, just keep the next newest one
+    List<UpdateEntity> updatesToDelete = selectionPolicy.selectUpdatesToDelete(Arrays.asList(updateDefault1, updateDefault2, updateRollout2), updateRollout2, manifestFilters);
+
+    Assert.assertEquals(1, updatesToDelete.size());
+    Assert.assertTrue(updatesToDelete.contains(updateDefault1));
+    Assert.assertFalse(updatesToDelete.contains(updateDefault2));
+    Assert.assertFalse(updatesToDelete.contains(updateRollout2));
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NormalCase_NewUpdate() {
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout2, updateRollout1, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NormalCase_NoUpdate() {
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout1, updateRollout1, manifestFilters);
+    Assert.assertFalse(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneMatchingFilters() {
+    // should choose to load an older update if the current update doesn't match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout1, updateDefault2, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NewerExists() {
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateRollout1, updateRollout2, manifestFilters);
+    Assert.assertFalse(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_OlderExistsMatchingFilters() {
+    // should not choose to load a new update that doesn't match the manifest filters
+    // if there is an existing older update that matches the filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, updateRollout1, manifestFilters);
+    Assert.assertFalse(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingCurrentFilters() {
+    // should choose to load a new update that doesn't match the manifest filters
+    // if no existing updates match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, null, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingNewFilters() {
+    // should choose to load a new update that doesn't match the manifest filters
+    // if no existing updates match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, updateDefault1, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testIsUpdateManifestFiltered_Nested() throws JSONException {
+    JSONObject nestedManifestFilters = new JSONObject("{\"nested.object.key\": \"different-value\"}");
```

Comment:
Using dots here precludes us from having dots inside of updateMetadata field names -- we need to update the spec if we are going to introduce this restriction.

Reply:
Good point. Discussed with @jkhales and we decided that we're just going to stick with non-nested key names for now, as this seems to cover all reasonable cases and is fully supported by the existing structured fields spec.